File: 2023/go/day08/src/main.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {

	data := getFileContent("src/input.txt")
	secquence, nodes := parseData(data)
	computePart1(secquence, nodes)

}

func getFileContent(path string) string {
	data, err := os.ReadFile(path)
	if err != nil {
		panic(err)
	}
	return string(data)
}

func parseData(data string) ([]int, map[string][]string) {
	sequence := make([]int, 0)
	nodes := make(map[string][]string)
	for index, line := range strings.Split(data, "\n") {
		if index == 0 {
			sequence = append(sequence, parseSequence(line)...)
			continue
		}
		if line == "" {
			continue
		}
		splited := strings.SplitN(line, " = ", 2)
		name, rest := splited[0], splited[1]
		leftRight := strings.Split(rest, ", ")
		left := strings.Replace(leftRight[0], "(", "", -1)
		right := strings.Replace(leftRight[1], ")", "", -1)
		value := make([]string, 2)
		value[0] = left
		value[1] = right
		nodes[name] = value
	}
	return sequence, nodes
}

func parseSequence(str string) []int {
	parsed := make([]int, len(str))
	for idx, r := range str {
		var i int
		if r == 'L' {
			i = 0
		} else {
			i = 1
		}
		parsed[idx] = i
	}
	return parsed
}

func computePart1(sequence []int, nodes map[string][]string) {
	current := "AAA"
	index := 0
	for current != "ZZZ" {
		current = nodes[current][sequence[index%len(sequence)]]
		index++
	}
	fmt.Printf("Result: %v\n", index)
}

File: 2023/rust/day08/src/main.rs
use std::collections::HashMap;

fn main() {
    let data = include_str!("./input.txt");
    let (sequence, nodes_string) = parse_data(data);
    compute_part1(sequence, nodes_string);
}

fn parse_data(data: &str) -> (Vec<usize>, HashMap<String, Vec<String>>) {
    let to_trim: &[_] = &[' ', '(', ')'];
    let mut split = data.split("\n");
    let sequence = split
        .next()
        .unwrap()
        .chars()
        .map(|c| if c == 'L' { 0 } else { 1 })
        .collect::<Vec<usize>>();
    let mut nodes = HashMap::new();
    for line in split.skip(1) {
        let mut tmp = Vec::new();
        let mut split = line.split("=");
        let key = String::from(split.next().unwrap().trim_end());
        let mut split = split.next().unwrap().split(", ");
        let left = String::from(split.next().unwrap().trim_matches(to_trim));
        let right = String::from(split.next().unwrap().trim_matches(to_trim));
        tmp.push(left);
        tmp.push(right);
        nodes.insert(key, tmp);
    }
    (sequence, nodes)
}

fn compute_part1(sequence: Vec<usize>, nodes: HashMap<String, Vec<String>>) {
    let mut current = "AAA".to_string();
    let mut index = 0;
    while current != "ZZZ".to_string() {
        current = nodes[&current][sequence[index % sequence.len()]].to_string();
        index += 1;
    }
    println!("index: {}", index);
}
